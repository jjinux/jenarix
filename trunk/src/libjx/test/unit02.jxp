
# unit02.jxp -- basic Python unit tests (no threads)

a = 1

print a

# we have literals 

b = {"some key":[1,2,3], 3:"some value", "another key":"another value",8.4:0}

###print b, len(b)

print len(b)

# REMEMBER, Jenarix uses copy semantics, not reference-taking semantics, so

d = {"some key":"value"}

c = d

c["another key"] ="value"

print c

print d

# is NOT how Python works, but it is how JXON works.

# we have print

print "hello Python world!!!" " Can you hear me?"

# we have for loops (simple iterator)

for b in range(3,8):
    print b

# we have while loops

x=1

while x<9:
    x=x+1
    print x

# does simple if work?

if 0:
    print "hi"
else:
    print 1234

# we have cascading IF statements

if 0:
    print "hi"
elif 0:
    print "who"
elif 1:
    print "wow"
else:
    print 1234

# we have functions

def fn(a,b):
    print "The sum of",a,"and",b,"is",a+b

fn(1,2)

fn(3,6)

# we have map / lambda

print map(lambda x:x*x, range(10))

# but we're still missing a lot of stuff...

# in particular...

# keyword arguments

# list comphrehensions

# nested loops

a=5
b=5
c=5
while a:
    a = a - 1
    b = a
    while b:
        c = b
        b = b - 1
        while c:
            c = c - 1
            print a,b,c
        while c:
            pass
        a = 0
    print b
    if 1:
       true
print b

c=2
print c
            
for a in [1,2]:
    for b in [5,6]:
        for c in [7,6]:
            print a,b,c

a=0
b=0
c=0
while a:
    while b:
        while c:
            pass

if 0:
    pass

# ordinary symbols

def fn():
    print [1,2,3]

a=fn

a()

del a

# this behavior is a little bizzare: calling an unbound (simple)
# symbol gives that symbol as the first member of an (unevalued) JXON
# invocation frame (symbol plus args)

assert([a,1,2], a(1,2))

# whereas with a compound symbol, we get null with the arguments

assert([null,3,4],a.b(3,4))

# entities and attributes

assert( entity(Base,null,[1,2,3],{x:"X",y:"Y"}), Base)

# "new" creates entity instances which have
# their own content, but inherit attributes

a=new(Base)
b=new(Base)

print a, b

print a.x, b.x

# setting attributes can be done through simple assignment

a.x=2

print a.x, b.x

print a,b

# instance content is accessed via the zero-length symbol (trailing dot)

print a.

# and can be acted upon using using builtin functions

print a..size()

# instance content can be assigned directly

a.=[2,3,4]

print a.

# and can be manipulated

append(a., 5)

print a.

a..reverse()

print a.

assert(4, a..take(1))

print a.

# derived entities can redefine attributes and include their own
# content prototype

assert( entity(Deriv, Base, [3,4,5], {x:"XX", z:"Z"}), Deriv)

a = new(Deriv)

print a.x, a.y, a.z

a.z="ZZZ"

print a.x, a.y, a.z

assert("ZZZ",a.z)

b = new(Deriv)

# notice how content is independent -- copied, not inherited

a..append(3)

print a., b.

# entity definitions with methods 

assert(
    entity(Base, null, [1,2,3], 
    { get_content: lambda self:self.,
      reverse_content: lambda self:self..reverse(),
      fill_content: lambda self,cnt,what: set(self., fill(cnt,what))
      }), Base )

a=new(Base)

print a.get_content()

a.=[4,5,6]

print a.get_content()

a.reverse_content()

print a.get_content()

a.fill_content(8,5)

print a.get_content()

del a

# one key thing to understand about jenarix is that dot notation
# cannot currently be used on intermedate results: a.b().d.e must
# instead be split up into two commands which involve binding of a
# temporary symbol which is then itself used with dot notation

a={b:lambda :{d:{e:[1,2]}}}

tmp = a.b()

print tmp.d.e

# this requirement arises from the Jenarix requirement that all
# resolved value references must refer to object which lie somewhere
# on the directed acyclic graph of the current scope (and are thus
# persistent)
