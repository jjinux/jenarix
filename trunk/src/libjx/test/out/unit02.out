# unit02.jxp -- basic Python unit tests (no threads)
# parsed: (set a 1)
# source: (set a 1)
#   eval: (`set a 1)
null;

# parsed: (print a)
# source: (print a)
#   eval: ( #native_fn`print (`resolve a))
1
null;

# we have literals 
# parsed: (set b {3:"some value", 8.4:0, "another key":"another value", "some key": [1, 2, 3]})
# source: (set b {3:"some value", 8.4:0, "another key":"another value", "some key": [1, 2, 3]})
#   eval: (`set b {3:"some value", 8.4:0, "another key":"another value", "some key": [1, 2, 3]})
null;

###print b, len(b)
# parsed: (print (len b))
# source: (print (size b))
#   eval: ( #native_fn`print (`size (`resolve b)))
4
null;

# REMEMBER, Jenarix uses copy semantics, not reference-taking semantics, so
# parsed: (set d {"some key":"value"})
# source: (set d {"some key":"value"})
#   eval: (`set d {"some key":"value"})
null;

# parsed: (set c d)
# source: (set c d)
#   eval: (`set c
#           (`resolve d))
null;

# parsed: (set (c (pass "another key")) "value")
# source: (set (c (pass "another key")) "value")
#   eval: (`set (c (`pass "another key")) "value")
null;

# parsed: (print c)
# source: (print c)
#   eval: ( #native_fn`print (`resolve c))
{"some key":"value","another key":"value"}
null;

# parsed: (print d)
# source: (print d)
#   eval: ( #native_fn`print (`resolve d))
{"some key":"value"}
null;

# is NOT how Python works, but it is how JXON works.
# we have print
# parsed: (print "hello Python world!!! Can you hear me?")
# source: (print "hello Python world!!! Can you hear me?")
#   eval: ( #native_fn`print "hello Python world!!! Can you hear me?")
hello Python world!!! Can you hear me?
null;

# we have for loops (simple iterator)
# we have while loops
# parsed: (foreach b
#           (range 3 8)
#           [ (print b)])
# source: (foreach b
#           (range 3 8)
#           [ (print b)])
#   eval: (`foreach b
#           (`range 3 8)
#           [ ( #native_fn`print (`resolve b))])
3
4
5
6
7
null;

# parsed: (set x 1)
# source: (set x 1)
#   eval: (`set x 1)
null;

# does simple if work?
# parsed: (while (lt x 9)
#           [ (set x
#               (add x 1)), (print x)])
# source: (while (lt x 9)
#           [ (set x
#               (add x 1)), (print x)])
#   eval: (`while (`lt (`resolve x) 9)
#           [ (`set x
#               (`add (`resolve x) 1)), ( #native_fn`print (`resolve x))])
2
3
4
5
6
7
8
9
null;

# we have cascading IF statements
# parsed: (if 0
#           [ (print "hi")]
#           [ (print 1234)])
# source: (if 0
#           [ (print "hi")]
#           [ (print 1234)])
#   eval: (`if 0
#           [ ( #native_fn`print "hi")]
#           [ ( #native_fn`print 1234)])
1234
null;

# we have functions
# parsed: (if 0
#           [ (print "hi")] 0
#           [ (print "who")] 1
#           [ (print "wow")]
#           [ (print 1234)])
# source: (if 0
#           [ (print "hi")] 0
#           [ (print "who")] 1
#           [ (print "wow")]
#           [ (print 1234)])
#   eval: (`if 0
#           [ ( #native_fn`print "hi")] 0
#           [ ( #native_fn`print "who")] 1
#           [ ( #native_fn`print "wow")]
#           [ ( #native_fn`print 1234)])
wow
null;

# parsed: (def fn
#           (a b)
#           [ (print "The sum of" a "and" b "is"
#               (add a b))])
# source: (def fn
#           (a b)
#           [ (print "The sum of" a "and" b "is"
#               (add a b))])
#   eval: (`def fn
#           (a b)
#           [ ( #native_fn`print "The sum of"
#               (`resolve a) "and"
#               (`resolve b) "is"
#               (`add (`resolve a)
#                 (`resolve b)))])
null;

# parsed: (fn 1 2)
# source: (fn 1 2)
#   eval: [ (`resolve fn),1,2]
The sum of 1 and 2 is 3
null;

# parsed: (fn 3 6)
# source: (fn 3 6)
#   eval: [ (`resolve fn),3,6]
The sum of 3 and 6 is 9
null;

# we have map / lambda
# parsed: (print (map (lambda (x)
#               (mul x x))
#             (range 10)))
# source: (print (map (lambda (x)
#               (mul x x))
#             (range 10)))
#   eval: ( #native_fn`print (`map (`lambda (x)
#               (`mul (`resolve x)
#                 (`resolve x)))
#             (`range 10)))
[0,1,4,9,16,25,36,49,64,81]
null;

# but we're still missing a lot of stuff...
# in particular...
# keyword arguments
# list comphrehensions
# nested loops
# parsed: (set a 5)
# source: (set a 5)
#   eval: (`set a 5)
null;

# parsed: (set b 5)
# source: (set b 5)
#   eval: (`set b 5)
null;

# parsed: (set c 5)
# source: (set c 5)
#   eval: (`set c 5)
null;

# parsed: (while a
#           [ (set a
#               (sub a 1)), (set b a),
#             (while b
#               [ (set c b), (set b
#                   (sub b 1)),
#                 (while c
#                   [ (set c
#                       (sub c 1)), (print a b c)]),
#                 (while c
#                   [ []]),
#                 (set a 0)]),
#             (print b),
#             (if 1
#               (true))])
# source: (while a
#           [ (set a
#               (sub a 1)), (set b a),
#             (while b
#               [ (set c b), (set b
#                   (sub b 1)),
#                 (while c
#                   [ (set c
#                       (sub c 1)), (print a b c)]),
#                 (while c
#                   [ []]),
#                 (set a 0)]),
#             (print b),
#             (if 1
#               [true])])
#   eval: (`while (`resolve a)
#           [ (`set a
#               (`sub (`resolve a) 1)), (`set b
#               (`resolve a)),
#             (`while (`resolve b)
#               [ (`set c
#                   (`resolve b)), (`set b
#                   (`sub (`resolve b) 1)),
#                 (`while (`resolve c)
#                   [ (`set c
#                       (`sub (`resolve c) 1)), ( #native_fn`print (`resolve a)
#                       (`resolve b)
#                       (`resolve c))]),
#                 (`while (`resolve c)
#                   [ []]),
#                 (`set a 0)]),
#             ( #native_fn`print (`resolve b)),
#             (`if 1
#               [true])])
4 3 3
4 3 2
4 3 1
4 3 0
0 2 2
0 2 1
0 2 0
0 1 1
0 1 0
0 0 0
0
true;

# parsed: (print b)
# source: (print b)
#   eval: ( #native_fn`print (`resolve b))
0
null;

# parsed: (set c 2)
# source: (set c 2)
#   eval: (`set c 2)
null;

# parsed: (print c)
# source: (print c)
#   eval: ( #native_fn`print (`resolve c))
2
null;

# parsed: (foreach a
#           [1, 2]
#           [ (foreach b
#               [5, 6]
#               [ (foreach c
#                   [7, 6]
#                   [ (print a b c)])])])
# source: (foreach a
#           [1, 2]
#           [ (foreach b
#               [5, 6]
#               [ (foreach c
#                   [7, 6]
#                   [ (print a b c)])])])
#   eval: (`foreach a
#           [1, 2]
#           [ (`foreach b
#               [5, 6]
#               [ (`foreach c
#                   [7, 6]
#                   [ ( #native_fn`print (`resolve a)
#                       (`resolve b)
#                       (`resolve c))])])])
1 5 7
1 5 6
1 6 7
1 6 6
2 5 7
2 5 6
2 6 7
2 6 6
null;

# parsed: (set a 0)
# source: (set a 0)
#   eval: (`set a 0)
null;

# parsed: (set b 0)
# source: (set b 0)
#   eval: (`set b 0)
null;

# parsed: (set c 0)
# source: (set c 0)
#   eval: (`set c 0)
null;

# parsed: (while a
#           [ (while b
#               [ (while c
#                   [ []])])])
# source: (while a
#           [ (while b
#               [ (while c
#                   [ []])])])
#   eval: (`while (`resolve a)
#           [ (`while (`resolve b)
#               [ (`while (`resolve c)
#                   [ []])])])
null;

# parsed: (if 0
#           [ []])
# source: (if 0
#           [ []])
#   eval: (`if 0
#           [ []])
null;

